cpmpy/expressions/utils.py

```diff
@@ -194,6 +194,17 @@ def get_bounds(expr):
             return int(expr), int(expr)
         return math.floor(expr), math.ceil(expr)
 
+def implies(expr, other):
+    """ like :func:`~cpmpy.expressions.core.Expression.implies`, but also safe to use for non-expressions """
+    if isinstance(expr, cp.expressions.core.Expression):
+        return expr.implies(other)
+    elif arg is True or arg is np.True_:
```

Comment:
This is the same logic as `is_true_cst` which is already in utils.
That being said, I don't think you need this helper function for this PR, and I prefer just checking whether `idx == i` is Boolean is (not) an expression in the Element decomposition.

We could introduce the helper-function nevertheless, but then we should also change the use of `.implies` in other places in the code base. I suppose we have this logic in multiple places at the moment, thereby explicitly checking for `is_bool` and converting to `BoolVal`.
 Better to open an separate issue and/or PR for that as it is out of scope for this one.

Reply:
Also, `arg` is undefined here, you mean `expr`

Reply:
the reason I don't use `is_true_cst` is because it would check more types than needed. Not sure if that is worth it though for readability.

Definitly good catch on the `arg`.

I thought this function will be helpfull in other casese, but I indeed didn't want to start looking for them in this PR. Please read the previous comment on the decomp itself, this function is necessary for it, if we change it I can put this logic inline and make a separate PR to fix this in some places.

Reply:
See new refactor of the decomp